src/basic/fs-util.c

```diff
@@ -359,22 +359,20 @@ int touch(const char *path) {
 }
 
 int symlink_idempotent(const char *from, const char *to) {
-        _cleanup_free_ char *p = NULL;
-        int r;
-
         assert(from);
         assert(to);
 
         if (symlink(from, to) < 0) {
+                _cleanup_free_ char *p = NULL;
+
                 if (errno != EEXIST)
                         return -errno;
 
-                r = readlink_malloc(to, &p);
-                if (r < 0)
-                        return r;
+                if (readlink_malloc(to, &p) < 0)
+                        return -EEXIST;
```

Comment:
I don't think this can go in with this commit message ;] Let's say that we get EIO here, because the disk is failing. I'd much rather see that propagated, then get a fake communication that the symlinks exists (and presumably points to a different target).

Frankly, if `readlink_malloc` fails, something is way off, and we shouldn't hide those errors.

Reply:
readlink_malloc() will fail with -EINVAL when the pre-existing file isn't actually a symlink. 

I force pushed a new version now, that propagates the readlink_malloc() error as is, except if it is EINVAL, which means, the file is not a symlink, in which case we propagate EEXIST as that's the original error.